### PR TITLE
[Feature] Add Decoupled KD Loss

### PR DESCRIPTION
**Pull request by spynccat**
## Motivation

Add Decoupled KD Loss (DKDLoss).
Add [DKDLoss](https://arxiv.org/pdf/2203.08679.pdf) config, readme and UT.
Add UT in `test_distillation_losses` for [DKDLoss](https://arxiv.org/pdf/2203.08679.pdf).

### CODE REVIEW

File: mmrazor/models/losses/decoupled_kd.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+    def _get_tckd_loss(
+        self,
+        student: torch.Tensor,
+        teacher: torch.Tensor,
+        gt_labels: torch.Tensor,
+        gt_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        """Calc target class knowledge distillation."""
+        other_mask = self._get_non_gt_mask(student, gt_labels)
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:97]: Maybe `non_gt_mask` is better? As the func name is `_get_non_gt_mask`.

File: mmrazor/models/losses/decoupled_kd.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        return kl_loss
+
+    def _cat_mask(
+        self,
+        target: torch.Tensor,
+        mask1: torch.Tensor,
+        mask2: torch.Tensor,
+    ) -> torch.Tensor:
+        t1 = (target * mask1).sum(dim=1, keepdims=True)
+        t2 = (target * mask2).sum(1, keepdims=True)
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:121]: `sum(dim=1, keepdims=True)`

File: mmrazor/models/losses/decoupled_kd.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            reduction=self.reduction) * self.tau**2
+        return kl_loss
+
+    def _cat_mask(
+        self,
+        target: torch.Tensor,
+        mask1: torch.Tensor,
+        mask2: torch.Tensor,
+    ) -> torch.Tensor:
+        t1 = (target * mask1).sum(dim=1, keepdims=True)
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:120]: It is better if the parameters' names make sense.

File: mmrazor/models/losses/decoupled_kd.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        teacher: torch.Tensor,
+    ) -> torch.Tensor:
+        kl_loss = F.kl_div(
+            student, teacher, size_average=False,
+            reduction=self.reduction) * self.tau**2
+        return kl_loss
+
+    def _cat_mask(
+        self,
+        target: torch.Tensor,
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:116]: It seems that `prob` is better than `target` as it could be the teacher's output probability or the student's one.

File: mmrazor/models/losses/decoupled_kd.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        t1 = (target * mask1).sum(dim=1, keepdims=True)
+        t2 = (target * mask2).sum(1, keepdims=True)
+        return torch.cat([t1, t2], dim=1)
+
+    def _get_gt_mask(
+        self,
+        logits: torch.Tensor,
+        target: torch.Tensor,
+    ) -> torch.Tensor:
+        target = target.reshape(-1)
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:129]: Pls add the docstring.

File: mmrazor/models/losses/decoupled_kd.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        return torch.zeros_like(logits).scatter_(1, target.unsqueeze(1),
+                                                 1).bool()
+
+    def _get_non_gt_mask(
+        self,
+        logits: torch.Tensor,
+        target: torch.Tensor,
+    ) -> torch.Tensor:
+        target = target.reshape(-1)
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:138]: Pls add the docstring.

File: mmrazor/models/losses/decoupled_kd.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        return self.loss_weight * loss
+
+    def _get_nckd_loss(
+        self,
+        student: torch.Tensor,
+        teacher: torch.Tensor,
+        gt_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        """Clac non-target class knowledge distillation."""
+        s_nckd = F.log_softmax(student / self.tau - 1000.0 * gt_mask, dim=1)
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:85]: Is the number 1000.0 a hyper-parameter? If so, maybe setting it as an attribute of `DKDLoss` is better.

File: mmrazor/models/losses/decoupled_kd.py (hunk 4 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py:116]: `Target` stands for the target class knowledge distillation (tckd) of teacher and student, not probe & target.

File: mmrazor/models/losses/decoupled_kd.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    def forward(
+        self,
+        student: torch.Tensor,
+        teacher: torch.Tensor,
+        data_samples: list,
+    ) -> torch.Tensor:
+        """DKDLoss forward function.
+
+        Args:
+            student (torch.Tensor): Student featuremap.
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: mmrazor/models/losses/decoupled_kd.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            f'but gets {reduction}.'
+        self.reduction = reduction
+        self.alpha = alpha
+        self.beta = beta
+        self.loss_weight = loss_weight
+
+    def forward(
+        self,
+        student: torch.Tensor,
+        teacher: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: configs/distill/mmcls/dkd/README.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+![avatar](../../../docs/imgs/model_zoo/dkd/dkd.png)
+
+## Results and models
```
[the remark this answers is not in the document]
Comment (spynccat) [reply configs/distill/mmcls/dkd/README.md:13]: fixed

File: mmrazor/models/losses/decoupled_kd.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        data_samples: list,
+    ) -> torch.Tensor:
+        """DKDLoss forward function.
+
+        Args:
+            student (torch.Tensor): Student featuremap.
+            teacher (torch.Tensor): Teacher featuremap.
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: mmrazor/models/losses/decoupled_kd.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            data_samples (list): List of model input sample data,
+            contains gt_label (dict of label and one-hot score)
+        """
+        # Unpack data samples and pack targets
+        if 'score' in data_samples[0].gt_label:
+            # Batch augmentation may convert labels to one-hot format scores.
+            gt_labels = torch.stack([i.gt_label.score for i in data_samples])
+        else:
+            gt_labels = torch.hstack([i.gt_label.label for i in data_samples])
```
[the remark this answers is not in the document]
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py:74]: Since distill head in gml is deprecated and replaced by `recorder`, `connector` and `loss_forward_mapping`, distill loss can only receive mapped params and cannot be hooked under existing framework.

File: mmrazor/models/losses/decoupled_kd.py (hunk 1 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: mmrazor/models/losses/decoupled_kd.py (hunk 2 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: mmrazor/models/losses/decoupled_kd.py (hunk 5 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py:144]: fixed

File: mmrazor/models/losses/decoupled_kd.py (hunk 6 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py:162]: fixed

File: mmrazor/models/losses/decoupled_kd.py (hunk 7 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: This is an implementation to mask out logits and is faster than index, please refer to [source code issue](https://github.com/megvii-research/mdistiller/issues/1).

File: mmrazor/models/losses/decoupled_kd.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            reduction=self.reduction) * self.tau**2
+        return kl_loss
+
+    def _cat_mask(
+        self,
+        tckd: torch.Tensor,
+        gt_mask: torch.Tensor,
+        non_gt_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        """Cat pt & pnt."""
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:125]: Pls replace `pt` and `pnt` with the full names.

File: mmrazor/models/losses/decoupled_kd.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        target: torch.Tensor,
+    ) -> torch.Tensor:
+        """Clac non-groundtruth mask on logits with target class tensor.
+
+        Args:
+            logits (torch.Tensor): The prediction logits with shape (N, C).
+            target (torch.Tensor): The gt_label target with shape (N, C).
+
+        Return:
+            torch.Tensor: The The maksed logits.
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:160]: Pls fix the typo.

File: mmrazor/models/losses/decoupled_kd.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        target: torch.Tensor,
+    ) -> torch.Tensor:
+        """Clac groundtruth mask on logits with target class tensor.
+
+        Args:
+            logits (torch.Tensor): The prediction logits with shape (N, C).
+            target (torch.Tensor): The gt_label target with shape (N, C).
+
+        Return:
+            torch.Tensor: The maksed logits.
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:142]: Pls fix the typo.

File: mmrazor/models/losses/decoupled_kd.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        mask_student = torch.log(self._cat_mask(s_tckd, gt_mask, non_gt_mask))
+        mask_teacher = self._cat_mask(t_tckd, gt_mask, non_gt_mask)
+        return self._kl_loss(mask_student, mask_teacher)
+
+    def _kl_loss(
+        self,
+        preds_S: torch.Tensor,
+        preds_T: torch.Tensor,
+    ) -> torch.Tensor:
+        """Calc kl."""
```
Comment (HIT-cwh) [inline mmrazor/models/losses/decoupled_kd.py:113]: Pls replace `Calc kl.` with `Calculate the KL Divergence.`

File: mmrazor/models/losses/decoupled_kd.py (hunk 7 above)
Comment (HIT-cwh) [reply mmrazor/models/losses/decoupled_kd.py]: Gotcha

File: mmrazor/models/losses/decoupled_kd.py (hunk 16 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: mmrazor/models/losses/decoupled_kd.py (hunk 15 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: mmrazor/models/losses/decoupled_kd.py (hunk 14 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: mmrazor/models/losses/decoupled_kd.py (hunk 13 above)
Comment (spynccat) [reply mmrazor/models/losses/decoupled_kd.py]: fixed

File: configs/distill/mmcls/dkd/README.md [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+https://mmclassification.readthedocs.io/en/latest/papers/resnet.html
+
+### Distillation training.
+
+```bash
+sh tools/slurm_train.sh $PARTITION $JOB_NAME \
+  configs/distill/mmcls/dkd/dkd_logits_r34_r18_8xb32_in1k.py \
+  $DISTILLATION_WORK_DIR
+```
```
Comment (FreakieHuang) [inline configs/distill/mmcls/dkd/README.md:53]: Add a special acknowledgement:
```
### Acknowledgement

Shout out to Davidgzx for his special contribution.

```
```
